File: backend/app/core/sentry.py
import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.starlette import StarletteIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from app.core.config import settings
import os

def init_sentry():
    """Initialize Sentry for error tracking"""
    # Only initialize in production
    if os.getenv("ENVIRONMENT", "development") == "production":
        sentry_sdk.init(
            dsn=os.getenv("SENTRY_DSN"),
            
            # Integrations
            integrations=[
                FastApiIntegration(
                    transaction_style="endpoint",
                    failed_request_status_codes={500, 502, 503, 504}
                ),
                StarletteIntegration(
                    transaction_style="endpoint"
                ),
                SqlalchemyIntegration(),
            ],
            
            # Performance monitoring
            traces_sample_rate=0.1,  # 10% of requests
            
            # Release tracking
            release=os.getenv("APP_VERSION", "unknown"),
            
            # Environment
            environment=os.getenv("ENVIRONMENT", "development"),
            
            # Don't send certain errors
            before_send=filter_events,
            
            # Attach additional data
            attach_stacktrace=True,
            send_default_pii=False,  # Don't send personally identifiable information
        )

def filter_events(event, hint):
    """Filter out certain events from being sent to Sentry"""
    if "exc_info" in hint:
        exc_type, exc_value, tb = hint["exc_info"]
        
        # Don't send client disconnections
        if exc_type.__name__ in ["ConnectionError", "ClientDisconnect"]:
            return None
            
        # Don't send validation errors
        if exc_type.__name__ == "RequestValidationError":
            return None
            
        # Don't send 404s
        if hasattr(exc_value, "status_code") and exc_value.status_code == 404:
            return None
    
    return event

File: frontend/src/sentry.js
import * as Sentry from "@sentry/react";

export function initSentry() {
  // Only initialize in production
  if (import.meta.env.PROD) {
    Sentry.init({
      dsn: import.meta.env.VITE_SENTRY_DSN,
      integrations: [
        Sentry.browserTracingIntegration(),
        Sentry.replayIntegration({
          maskAllText: true,
          blockAllMedia: true,
        }),
      ],
      // Performance Monitoring
      tracesSampleRate: 0.1, // 10% of transactions
      
      // Session Replay
      replaysSessionSampleRate: 0.1, // 10% of sessions
      replaysOnErrorSampleRate: 1.0, // 100% of sessions with errors
      
      // Release tracking
      release: import.meta.env.VITE_APP_VERSION || "unknown",
      
      // Environment
      environment: import.meta.env.MODE,
      
      // Don't send errors in development
      beforeSend(event, hint) {
        // Filter out certain errors
        const error = hint.originalException;
        
        // Don't send network errors
        if (error?.message?.includes('Network request failed')) {
          return null;
        }
        
        // Don't send authentication errors
        if (error?.message?.includes('401') || error?.message?.includes('403')) {
          return null;
        }
        
        return event;
      },
    });
  }
}